opencti-platform/opencti-graphql/src/database/middleware.js

```diff
@@ -1962,7 +1963,24 @@ export const updateAttribute = async (context, user, id, type, inputs, opts = {}
         if (relType === RELATION_GRANTED_TO && !userHaveCapability(user, KNOWLEDGE_ORGANIZATION_RESTRICT)) {
           throw ForbiddenAccess();
         }
-        const { value: refs, operation = UPDATE_OPERATION_REPLACE } = meta[metaIndex];
+
+        let { value: refs, operation = UPDATE_OPERATION_REPLACE } = meta[metaIndex];
+
+        const currentMarkings = initial.objectMarking;
+        if (relType === RELATION_OBJECT_MARKING && currentMarkings) {
+          const markingsMap = await getEntitiesMapFromCache(context, SYSTEM_USER, ENTITY_TYPE_MARKING_DEFINITION);
+          const markingAdded = markingsMap.get(R.head(refs));
+
+          const markingsHasSameType = currentMarkings.filter((currentMarking) => currentMarking.definition_type === markingAdded.definition_type);
+          const markingsToReplace = markingsHasSameType.filter((currentMarking) => currentMarking.x_opencti_order !== markingAdded.x_opencti_order);
+          if (markingsToReplace.length !== 0) {
+            // filter every value that has not the same type
+            const existingMarkings = currentMarkings.filter((marking) => !marking.definition_type.includes(markingAdded.definition_type)).map((m) => m.id);
+            const newMarkings = existingMarkings.concat(markingAdded.id);
+            ({ operation, refs } = { operation: UPDATE_OPERATION_REPLACE, refs: newMarkings });
```

Comment:
Can we move some logic inside the markingDefinition-utils ?

To separate the concepts and to have the management of sorting/sorting/filtering of markings in the domain and the mechanism of compute operations in middleware?

Reply:
Yes, done :)